// angular-admin-lte/esm5/lib/color/color.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, Renderer2, ElementRef } from '@angular/core';
import { ColorService } from './color.service';
var BackgroundColorDirective = /** @class */ (function () {
    /**
     * @method constructor
     * @param elementRef   [description]
     * @param renderer2    [description]
     * @param colorService [description]
     */
    function BackgroundColorDirective(elementRef, renderer2, colorService) {
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        this.colorService = colorService;
        this.condition = true;
    }
    Object.defineProperty(BackgroundColorDirective.prototype, "setPrefix", {
        set: function (prefix) {
            this.prefix = prefix;
            this.colorService.setBackgroundColor(this.color, this.condition, this.property, this.prefix);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BackgroundColorDirective.prototype, "setColor", {
        set: function (color) {
            this.color = color;
            this.colorService.setBackgroundColor(this.color, this.condition, this.property, this.prefix);
        },
        enumerable: true,
        configurable: true
    });
    BackgroundColorDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: ColorService }
    ]; };
    __decorate([
        Input('mkColorCondition'),
        __metadata("design:type", Object)
    ], BackgroundColorDirective.prototype, "condition", void 0);
    __decorate([
        Input('mkColorPrefix'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], BackgroundColorDirective.prototype, "setPrefix", null);
    __decorate([
        Input('mkColorProperty'),
        __metadata("design:type", String)
    ], BackgroundColorDirective.prototype, "property", void 0);
    __decorate([
        Input('mkColor'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], BackgroundColorDirective.prototype, "setColor", null);
    BackgroundColorDirective = __decorate([
        Directive({
            selector: '[mkColor]',
            providers: [ColorService]
        }),
        __metadata("design:paramtypes", [ElementRef,
            Renderer2,
            ColorService])
    ], BackgroundColorDirective);
    return BackgroundColorDirective;
}());
export { BackgroundColorDirective };
var ColorDirective = /** @class */ (function () {
    /**
     * @method constructor
     * @param elementRef   [description]
     * @param renderer2    [description]
     * @param colorService [description]
     */
    function ColorDirective(elementRef, renderer2, colorService) {
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        this.colorService = colorService;
    }
    Object.defineProperty(ColorDirective.prototype, "color", {
        set: function (color) {
            this.colorService.setFontColor(color);
        },
        enumerable: true,
        configurable: true
    });
    ColorDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: ColorService }
    ]; };
    __decorate([
        Input('mkFontColor'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], ColorDirective.prototype, "color", null);
    ColorDirective = __decorate([
        Directive({
            selector: '[mkFontColor]',
            providers: [ColorService]
        }),
        __metadata("design:paramtypes", [ElementRef,
            Renderer2,
            ColorService])
    ], ColorDirective);
    return ColorDirective;
}());
export { ColorDirective };
