// angular-admin-lte/esm5/lib/box-info/box-info.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ContentChild } from '@angular/core';
import { BoxInfoContentDirective, BoxInfoFooterDirective, BoxInfoHeaderDirective } from './box-info.directive';
/*
 *
 */
var BoxInfoComponent = /** @class */ (function () {
    function BoxInfoComponent() {
        this.contentStyleClass = 'info-box-number';
        this.footerStyleClass = 'progress-description';
        this.headerStyleClass = 'info-box-text';
        this.iconColor = '#fff';
        this.iconStyleClass = 'ion ion-bag';
        this.styleClass = 'info-box';
    }
    /**
     * @method ngOnInit
     */
    BoxInfoComponent.prototype.ngOnInit = function () {
        if (!this.backgroundColor) {
            this.progressBarBg = this.iconBackgroundColor;
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxInfoComponent.prototype, "backgroundColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxInfoComponent.prototype, "contentStyleClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxInfoComponent.prototype, "contentColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxInfoComponent.prototype, "footer", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxInfoComponent.prototype, "footerColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxInfoComponent.prototype, "footerStyleClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxInfoComponent.prototype, "header", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxInfoComponent.prototype, "headerColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxInfoComponent.prototype, "headerStyleClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxInfoComponent.prototype, "iconBackgroundColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxInfoComponent.prototype, "iconColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxInfoComponent.prototype, "iconStyleClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], BoxInfoComponent.prototype, "progressWidth", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxInfoComponent.prototype, "styleClass", void 0);
    __decorate([
        ContentChild(BoxInfoHeaderDirective, /* TODO: add static flag */ {}),
        __metadata("design:type", BoxInfoHeaderDirective)
    ], BoxInfoComponent.prototype, "boxInfoHeaderDirective", void 0);
    __decorate([
        ContentChild(BoxInfoFooterDirective, /* TODO: add static flag */ {}),
        __metadata("design:type", BoxInfoFooterDirective)
    ], BoxInfoComponent.prototype, "boxInfoFooterDirective", void 0);
    __decorate([
        ContentChild(BoxInfoContentDirective, /* TODO: add static flag */ {}),
        __metadata("design:type", BoxInfoContentDirective)
    ], BoxInfoComponent.prototype, "boxInfoContentDirective", void 0);
    BoxInfoComponent = __decorate([
        Component({
            selector: 'mk-box-info',
            template: "<div [ngClass]=\"styleClass\" [mkColor]=\"backgroundColor\" mkColorProperty=\"background-color\">\n  <span class=\"info-box-icon\" [mkColor]=\"iconBackgroundColor\" mkColorProperty=\"background-color\">\n    <i [ngClass]=\"iconStyleClass\" [mkFontColor]=\"iconColor\"></i>\n  </span>\n  <div class=\"info-box-content\">\n    <span *ngIf=\"header || boxInfoHeaderDirective\" [ngClass]=\"headerStyleClass\" [mkFontColor]=\"headerColor\">\n      {{header}}\n      <ng-content select=\"mk-box-header\"></ng-content>\n    </span>\n    <span [ngClass]=\"contentStyleClass\" [mkFontColor]=\"contentColor\">\n      <ng-container *ngIf=\"boxInfoHeaderDirective || boxInfoContentDirective || boxInfoFooterDirective; else noDirective\">\n        <ng-content select=\"mk-box-content\"></ng-content>\n      </ng-container>\n      <ng-template #noDirective>\n        <ng-content></ng-content>\n      </ng-template>\n    </span>\n    <div *ngIf=\"progressWidth\" class=\"progress\">\n      <div class=\"progress-bar\" [mkColor]=\"progressBarBg\" mkColorProperty=\"background-color\" [style.width.%]=\"progressWidth\"></div>\n    </div>\n    <span *ngIf=\"footer || boxInfoFooterDirective\" [ngClass]=\"footerStyleClass\" [mkFontColor]=\"footerColor\">\n      {{footer}}\n      <ng-content select=\"mk-box-footer\"></ng-content>\n    </span>\n  </div>\n</div>\n",
            styles: [".info-box.bg-color>.info-box-content{color:#fff}"]
        })
    ], BoxInfoComponent);
    return BoxInfoComponent;
}());
export { BoxInfoComponent };
