// angular-admin-lte/esm5/lib/layout/footer/footer.component.js
import { __decorate, __metadata } from "tslib";
import { Component, ContentChild, OnInit, ViewChild, TemplateRef, ElementRef } from '@angular/core';
import { FooterService } from './footer.service';
/**
 * Footer Left
 */
var FooterLeftComponent = /** @class */ (function () {
    function FooterLeftComponent() {
    }
    __decorate([
        ViewChild('templateRef', { static: true }),
        __metadata("design:type", TemplateRef)
    ], FooterLeftComponent.prototype, "templateRef", void 0);
    FooterLeftComponent = __decorate([
        Component({
            selector: 'mk-layout-footer-left',
            template: '<ng-template #templateRef><ng-content></ng-content></ng-template>'
        })
    ], FooterLeftComponent);
    return FooterLeftComponent;
}());
export { FooterLeftComponent };
/**
 * Footer Right
 */
var FooterRightComponent = /** @class */ (function () {
    function FooterRightComponent() {
    }
    __decorate([
        ViewChild('templateRef', { static: true }),
        __metadata("design:type", TemplateRef)
    ], FooterRightComponent.prototype, "templateRef", void 0);
    FooterRightComponent = __decorate([
        Component({
            selector: 'mk-layout-footer-right',
            template: '<ng-template #templateRef><ng-content></ng-content></ng-template>'
        })
    ], FooterRightComponent);
    return FooterRightComponent;
}());
export { FooterRightComponent };
var FooterComponent = /** @class */ (function () {
    function FooterComponent(elementRef, footerService) {
        this.elementRef = elementRef;
        this.footerService = footerService;
    }
    FooterComponent.prototype.ngOnInit = function () {
        this.footerService.elementRef = this.elementRef;
    };
    FooterComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: FooterService }
    ]; };
    __decorate([
        ContentChild(FooterLeftComponent, /* TODO: add static flag */ {}),
        __metadata("design:type", FooterLeftComponent)
    ], FooterComponent.prototype, "footerLeftComponent", void 0);
    __decorate([
        ContentChild(FooterRightComponent, /* TODO: add static flag */ {}),
        __metadata("design:type", FooterRightComponent)
    ], FooterComponent.prototype, "footerRightComponent", void 0);
    FooterComponent = __decorate([
        Component({
            selector: 'mk-layout-footer',
            template: "<footer class=\"main-footer\">\n  <div class=\"pull-right hidden-xs\">\n    <ng-template [ngTemplateOutlet]=\"footerRightComponent?.templateRef\"></ng-template>\n  </div>\n  <ng-template [ngTemplateOutlet]=\"footerLeftComponent?.templateRef\"></ng-template>\n</footer>\n",
            styles: [":host{display:block}"]
        }),
        __metadata("design:paramtypes", [ElementRef,
            FooterService])
    ], FooterComponent);
    return FooterComponent;
}());
export { FooterComponent };
