// angular-admin-lte/esm5/lib/box-small/box-small.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ContentChild } from '@angular/core';
import { BoxSmallContentDirective, BoxSmallFooterDirective, BoxSmallHeaderDirective } from './box-small.directive';
/*
 *
 */
var BoxSmallComponent = /** @class */ (function () {
    function BoxSmallComponent() {
        this.contentStyleClass = 'small-box-content';
        this.footerStyleClass = 'small-box-footer';
        this.headerStyleClass = 'small-box-header';
        this.iconStyleClass = 'ion ion-bag';
        this.styleClass = 'small-box';
    }
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxSmallComponent.prototype, "backgroundColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxSmallComponent.prototype, "contentColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxSmallComponent.prototype, "contentStyleClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxSmallComponent.prototype, "footer", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxSmallComponent.prototype, "footerColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxSmallComponent.prototype, "footerStyleClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxSmallComponent.prototype, "header", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxSmallComponent.prototype, "headerColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxSmallComponent.prototype, "headerStyleClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], BoxSmallComponent.prototype, "iconColor", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxSmallComponent.prototype, "iconStyleClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], BoxSmallComponent.prototype, "styleClass", void 0);
    __decorate([
        ContentChild(BoxSmallHeaderDirective, /* TODO: add static flag */ {}),
        __metadata("design:type", BoxSmallHeaderDirective)
    ], BoxSmallComponent.prototype, "boxSmallHeaderDirective", void 0);
    __decorate([
        ContentChild(BoxSmallFooterDirective, /* TODO: add static flag */ {}),
        __metadata("design:type", BoxSmallFooterDirective)
    ], BoxSmallComponent.prototype, "boxSmallFooterDirective", void 0);
    __decorate([
        ContentChild(BoxSmallContentDirective, /* TODO: add static flag */ {}),
        __metadata("design:type", BoxSmallContentDirective)
    ], BoxSmallComponent.prototype, "boxSmallContentDirective", void 0);
    BoxSmallComponent = __decorate([
        Component({
            selector: 'mk-box-small',
            template: "<div [ngClass]=\"styleClass\" [mkColor]=\"backgroundColor\" mkColorProperty=\"background-color\">\n  <div class=\"inner\">\n    <h3 *ngIf=\"header || boxSmallHeaderDirective\" [ngClass]=\"headerStyleClass\" [mkFontColor]=\"headerColor\">\n      {{header}}\n      <ng-content select=\"mk-box-small-header\"></ng-content>\n    </h3>\n    <p [ngClass]=\"contentStyleClass\" [mkFontColor]=\"contentColor\">\n      <ng-container *ngIf=\"boxSmallHeaderDirective || boxSmallContentDirective || boxSmallFooterDirective; else noDirective\">\n        <ng-content select=\"mk-box-small-content\"></ng-content>\n      </ng-container>\n      <ng-template #noDirective>\n        <ng-content></ng-content>\n      </ng-template>\n    </p>\n  </div>\n  <div *ngIf=\"iconStyleClass\" class=\"icon\">\n    <i [ngClass]=\"iconStyleClass\" [mkFontColor]=\"iconColor\"></i>\n  </div>\n  <span *ngIf=\"footer || boxSmallFooterDirective\" [ngClass]=\"footerStyleClass\" [mkFontColor]=\"footerColor\">\n    {{footer}}\n    <ng-content select=\"mk-box-small-footer\"></ng-content>\n  </span>\n</div>\n",
            styles: [".small-box.bg-color{color:#fff}/deep/ .small-box-footer:hover{cursor:pointer}/deep/ .small-box-footer a{color:rgba(255,255,255,.8)}/deep/ .small-box-footer:hover a{color:#fff}"]
        })
    ], BoxSmallComponent);
    return BoxSmallComponent;
}());
export { BoxSmallComponent };
