// angular-admin-lte/esm2015/lib/layout/footer/footer.component.js
import { __decorate, __metadata } from "tslib";
import { Component, ContentChild, OnInit, ViewChild, TemplateRef, ElementRef } from '@angular/core';
import { FooterService } from './footer.service';
/**
 * Footer Left
 */
let FooterLeftComponent = class FooterLeftComponent {
};
__decorate([
    ViewChild('templateRef', { static: true }),
    __metadata("design:type", TemplateRef)
], FooterLeftComponent.prototype, "templateRef", void 0);
FooterLeftComponent = __decorate([
    Component({
        selector: 'mk-layout-footer-left',
        template: '<ng-template #templateRef><ng-content></ng-content></ng-template>'
    })
], FooterLeftComponent);
export { FooterLeftComponent };
/**
 * Footer Right
 */
let FooterRightComponent = class FooterRightComponent {
};
__decorate([
    ViewChild('templateRef', { static: true }),
    __metadata("design:type", TemplateRef)
], FooterRightComponent.prototype, "templateRef", void 0);
FooterRightComponent = __decorate([
    Component({
        selector: 'mk-layout-footer-right',
        template: '<ng-template #templateRef><ng-content></ng-content></ng-template>'
    })
], FooterRightComponent);
export { FooterRightComponent };
let FooterComponent = class FooterComponent {
    constructor(elementRef, footerService) {
        this.elementRef = elementRef;
        this.footerService = footerService;
    }
    ngOnInit() {
        this.footerService.elementRef = this.elementRef;
    }
};
FooterComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: FooterService }
];
__decorate([
    ContentChild(FooterLeftComponent, /* TODO: add static flag */ {}),
    __metadata("design:type", FooterLeftComponent)
], FooterComponent.prototype, "footerLeftComponent", void 0);
__decorate([
    ContentChild(FooterRightComponent, /* TODO: add static flag */ {}),
    __metadata("design:type", FooterRightComponent)
], FooterComponent.prototype, "footerRightComponent", void 0);
FooterComponent = __decorate([
    Component({
        selector: 'mk-layout-footer',
        template: "<footer class=\"main-footer\">\n  <div class=\"pull-right hidden-xs\">\n    <ng-template [ngTemplateOutlet]=\"footerRightComponent?.templateRef\"></ng-template>\n  </div>\n  <ng-template [ngTemplateOutlet]=\"footerLeftComponent?.templateRef\"></ng-template>\n</footer>\n",
        styles: [":host{display:block}"]
    }),
    __metadata("design:paramtypes", [ElementRef,
        FooterService])
], FooterComponent);
export { FooterComponent };
