// angular-admin-lte/esm2015/lib/form/input-group/input-group.component.js
import { __decorate, __metadata } from "tslib";
import { Component, ContentChild, Input } from '@angular/core';
import { removeSubscriptions } from '../../helpers';
import { InputGroupAddonLeftDirective, InputGroupAddonRightDirective, InputGroupContentDirective, InputGroupLabelDirective } from './input-group.directive';
import { InputTextDirective } from '../input-text/input-text.directive';
/*
 *
 */
let InputGroupComponent = class InputGroupComponent {
    constructor() {
        this.subscriptions = [];
        this.inputColor = 'default';
        this.inputErrorColor = 'danger';
        this.inputValidColor = 'success';
        this.wrapperClasses = 'form-group';
    }
    ngAfterContentInit() {
        this.subscriptions.push(this.inputTextDirective.onKeyup.subscribe((value) => {
            if (value.invalid) {
                this.currentColor = this.inputErrorColor;
                this.currentFontColor = this.inputErrorFontColor;
            }
            else if (!value.invalid) {
                this.currentColor = this.inputValidColor;
                this.currentFontColor = this.inputValidFontColor;
            }
            else {
                this.currentColor = this.inputColor;
                this.currentFontColor = this.inputFontColor;
            }
        }));
    }
    ngOnDestroy() {
        removeSubscriptions(this.subscriptions);
    }
};
__decorate([
    Input(),
    __metadata("design:type", String)
], InputGroupComponent.prototype, "addonLeft", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], InputGroupComponent.prototype, "addonRight", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], InputGroupComponent.prototype, "inputColor", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], InputGroupComponent.prototype, "inputFontColor", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], InputGroupComponent.prototype, "inputErrorColor", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], InputGroupComponent.prototype, "inputErrorFontColor", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], InputGroupComponent.prototype, "inputValidColor", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], InputGroupComponent.prototype, "inputValidFontColor", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], InputGroupComponent.prototype, "label", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], InputGroupComponent.prototype, "wrapperClasses", void 0);
__decorate([
    ContentChild(InputGroupLabelDirective),
    __metadata("design:type", InputGroupLabelDirective)
], InputGroupComponent.prototype, "inputGroupLabelDirective", void 0);
__decorate([
    ContentChild(InputGroupAddonLeftDirective),
    __metadata("design:type", InputGroupAddonLeftDirective)
], InputGroupComponent.prototype, "inputGroupAddonLeftDirective", void 0);
__decorate([
    ContentChild(InputGroupAddonRightDirective),
    __metadata("design:type", InputGroupAddonRightDirective)
], InputGroupComponent.prototype, "inputGroupAddonRightDirective", void 0);
__decorate([
    ContentChild(InputGroupContentDirective),
    __metadata("design:type", InputGroupContentDirective)
], InputGroupComponent.prototype, "inputGroupContentDirective", void 0);
__decorate([
    ContentChild(InputTextDirective),
    __metadata("design:type", InputTextDirective)
], InputGroupComponent.prototype, "inputTextDirective", void 0);
InputGroupComponent = __decorate([
    Component({
        selector: 'mk-input-group',
        template: "<div [ngClass]=\"wrapperClasses\" [mkColor]=\"currentColor || inputColor\" mkColorPrefix=\"has\">\n  <label *ngIf=\"label || inputGroupLabelDirective\">\n    {{label}}\n    <ng-content select=\"mk-input-group-label\"></ng-content>\n  </label>\n  <div *ngIf=\"addonLeft || inputGroupAddonLeftDirective || addonRight || inputGroupAddonRightDirective; else noAddon\" class=\"input-group\">\n    <span *ngIf=\"addonLeft || inputGroupAddonLeftDirective\" class=\"input-group-addon\">\n      {{addonLeft}}\n      <ng-content select=\"mk-input-group-addon-left\"></ng-content>\n    </span>\n    <ng-content select=\"mk-input-group-content\"></ng-content>\n    <span *ngIf=\"addonRight || inputGroupAddonRightDirective\" class=\"input-group-addon\">\n      {{addonRight}}\n      <ng-content select=\"mk-input-group-addon-right\"></ng-content>\n    </span>\n  </div>\n  <ng-template #noAddon><ng-content select=\"mk-input-group-content\"></ng-content></ng-template>\n</div>\n"
    })
], InputGroupComponent);
export { InputGroupComponent };
